src/lib/tensor.js

```diff
@@ -28,7 +28,8 @@ export class Tensor {
       // Copy the data.
       this.data = data.slice();
     } else {
-      this.data = new Array(size).fill(0);
+      // cast float64 to float32
+      this.data = new Float32Array(size).fill(0);
     }
```

Comment:
For float16, we'll need to do the same. Should the tensor just take a data type parameter? Otherwise we won't be able to distinguish between float16 and float32.

However, I don't see a Float16Array in ES 🤔. Well it should okay to use Float32Array and zero the bottom 13 bits (IEEE float16 has 10 fraction bits whereas float32 has 23 fraction bits).

Reply:
Chai commented previously 

> Although the lack of FP16 support in the ArrayBufferView could possibly be worked around by just using Uint16Array to appropriately offset the tensor data and deal with the float-casting inside the webnn implementation itself, it could still be confusing to some at the model builder level.

 Workaround by 森建 on https://esdiscuss.org/topic/float16array

> WebGL 2.0 supports half-precision float (16bit float) by default.
> But now, we must use the following dirty hack using Uint16Array.


```js
// ref: http://stackoverflow.com/questions/32633585/how-do-you-convert-to-half-floats-in-javascript
var toHalf = (function() {

   var floatView = new Float32Array(1);
   var int32View = new Int32Array(floatView.buffer);

   /* This method is faster than the OpenEXR implementation (very often
    * used, eg. in Ogre), with the additional benefit of rounding, inspired
    * by James Tursa?s half-precision code. */
   return function toHalf(val) {

     floatView[0] = val;
     var x = int32View[0];

     var bits = (x >> 16) & 0x8000; /* Get the sign */
     var m = (x >> 12) & 0x07ff; /* Keep one extra bit for rounding */
     var e = (x >> 23) & 0xff; /* Using int is faster here */

     /* If zero, or denormal, or exponent underflows too much for a denormal
      * half, return signed zero. */
     if (e < 103) {
       return bits;
     }

     /* If NaN, return NaN. If Inf or exponent overflow, return Inf. */
     if (e > 142) {
       bits |= 0x7c00;
       /* If exponent was 0xff and one mantissa bit was set, it means NaN,
        * not Inf, so make sure we set one mantissa bit too. */
       bits |= ((e == 255) ? 0 : 1) && (x & 0x007fffff);
       return bits;
     }

     /* If exponent underflows but not too much, return a denormal */
     if (e < 113) {
       m |= 0x0800;
       /* Extra rounding may overflow and set mantissa to 0 and exponent
        * to 1, which is OK. */
       bits |= (m >> (114 - e)) + ((m >> (113 - e)) & 1);
       return bits;
     }

     bits |= ((e - 112) << 10) | (m >> 1);
     /* Extra rounding. An overflow will set mantissa to 0 and increment
      * the exponent, which is OK. */
     bits += m & 1;
     return bits;
   };

}());

var tex = new Uint16Array(4);
tex[0] = toHalf(0.5);
tex[1] = toHalf(1);
tex[2] = toHalf(123);
tex[3] = toHalf(-13);
```

@fdwr  Please have a look firstly, I will check above workaround.

Reply:
What about the double-precision? Should the `Float64Array` be used here?

Reply:
Yes, you are correct. 
`Number` type is default `float64` in JavaScript, using `Float64Array` will be explicit.
I will submit a PR to support this feature.

Reply:
Thanks for the update. Could we introduce a helper similar to [`float64ToFloat32()`](https://github.com/math-io/float64-to-float32/blob/master/lib/index.js#L17)? With that, you don't need to set the array type of tensor, you can just use this helper method to truncate the inputs and outputs to float32. 

What do you think?

Reply:
My concern is if the `data` is a typed array, say `Float32Array`, V8 may generate the single precision instruction to calculate them, say `tensor_c.data[i] = tensor_a.data[i] + tensor_b.data[2]` if `data` is `Float32Array`. My proposal is to keep the ref baseline implementation double precision (by leaving the `tensor.data` to use JS Array) and just truncate the inputs and outputs to float32. You can use either `Float32Array` to convert a whole array or `float64ToFloat32()` helper to convert a single number.

Reply:
> My concern is if the data is a typed array...

That makes sense. You only want the input and output to be truncated to float32, but all the intermediate calculations would/should be float64.

Reply:
Thanks @fdwr and @huningxin for your insightful inputs, I will follow this solution.

Reply:
I just noticed this is still open 😳 . I'm out of town this week, but will re-review upon returning... ⏳

Reply:
Thanks for soft reminder. My those PRs have been submitted onto https://github.com/webmachinelearning/webnn-baseline/pulls?q=is%3Apr+is%3Aopen+generate which would be updated later after I focused on implementing WPT tests of remainder ops. 

So I would close this PR here. Thanks